File: shared/contracts/skill-record.ts
export type SkillRecordKind = 'build' | 'runtime';

export type SkillRecordStatus = 'active' | 'draft' | 'deprecated';

export type SkillRecordOperator = 'gt' | 'lt' | 'gte' | 'lte' | 'eq';

export interface SkillRecordTrigger {
  failure_family?: string | null;
  metric_name?: string | null;
  threshold?: number | null;
  operator?: SkillRecordOperator;
  blame_pattern?: string | null;
}

export interface SkillRecordEvalCriterion {
  metric: string;
  target: number;
  operator: SkillRecordOperator;
  weight: number;
}

export interface SkillRecordToolDefinition {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  returns?: Record<string, unknown> | null;
  implementation?: string | null;
  sandbox_policy?: string;
}

export interface SkillRecordPolicy {
  name: string;
  description: string;
  rule_type: string;
  condition: string;
  action: string;
  severity?: string;
}

export interface SkillRecordTestCase {
  name: string;
  description: string;
  input: Record<string, unknown>;
  expected_output?: Record<string, unknown> | null;
  expected_behavior?: string | null;
  assertions: string[];
}

export interface SkillRecordEffectiveness {
  times_applied: number;
  success_count: number;
  success_rate: number;
  avg_improvement: number;
  total_improvement: number;
  last_applied?: number | null;
}

export interface SkillRecordDependency {
  skill_id: string;
  version_constraint: string;
  optional: boolean;
}

export interface SkillRecord {
  skill_id: string;
  name: string;
  kind: SkillRecordKind;
  version: string;
  domain: string;
  description: string;
  status: SkillRecordStatus;
  tags: string[];
  triggers: SkillRecordTrigger[];
  eval_criteria: SkillRecordEvalCriterion[];
  tools: SkillRecordToolDefinition[];
  policies: SkillRecordPolicy[];
  test_cases: SkillRecordTestCase[];
  dependencies: SkillRecordDependency[];
  effectiveness: SkillRecordEffectiveness;
  created_at: string;
  updated_at: string;
  metadata?: Record<string, unknown>;
}

File: shared/contracts/skill_record.py
"""Shared skill record contract."""

from __future__ import annotations

from dataclasses import dataclass, asdict, field
from typing import Any


@dataclass(slots=True)
class SkillRecord:
    """Describe a durable skill entry shared by CLI, API, and UI."""

    skill_id: str
    name: str
    kind: str
    version: str
    domain: str
    status: str
    description: str = ""
    tags: list[str] = field(default_factory=list)
    effectiveness: dict[str, Any] = field(default_factory=dict)
    source: str = ""
    skill_layer: str = ""  # "build", "runtime", or "" (unclassified)
    created_at: str = ""
    updated_at: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Return a JSON-safe representation for persistence and transport."""
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> SkillRecord:
        """Rehydrate a skill record from persisted data."""
        return cls(
            skill_id=data["skill_id"],
            name=data["name"],
            kind=data["kind"],
            version=data["version"],
            domain=data["domain"],
            status=data["status"],
            description=data.get("description", ""),
            tags=list(data.get("tags", [])),
            effectiveness=dict(data.get("effectiveness", {})),
            source=data.get("source", ""),
            skill_layer=data.get("skill_layer", ""),
            created_at=data.get("created_at", ""),
            updated_at=data.get("updated_at", ""),
            metadata=dict(data.get("metadata", {})),
        )
